flex-array/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+                type Item = T;
+                type IntoIter = IntoIter<T, $len>;
+
+                /// Creates a consuming iterator, that is, one that moves each value out of
+                /// the array (from start to end). The array cannot be used after calling
+                /// this unless `T` implements `Copy`, so the whole array is copied.
+                fn into_iter(self) -> Self::IntoIter {
+                    self.0.into_iter()
+                }
+            }
+
+            impl<'a, T> IntoIterator for &'a Array<T, typenum::$ty> {
+                type Item = &'a T;
+                type IntoIter = Iter<'a, T>;
+
+                fn into_iter(self) -> Iter<'a, T> {
+                    self.iter()
+                }
+            }
+
+            impl<'a, T> IntoIterator for &'a mut Array<T, typenum::$ty> {
+                type Item = &'a mut T;
+                type IntoIter = IterMut<'a, T>;
+
+                #[inline]
+                fn into_iter(self) -> IterMut<'a, T> {
+                    self.iter_mut()
+                }
+            }
+
+            impl<T> TryFrom<&[T]> for Array<T, typenum::$ty>
+            where
+                T: Copy,
+            {
+                type Error = TryFromSliceError;
+
+                #[inline]
+                fn try_from(slice: &[T]) -> Result<Array<T, typenum::$ty>, TryFromSliceError> {
+                    Self::from_slice(slice)
+                }
+            }
+        )+
+     };
+}
+
+impl_array_length! {
+    0 => U0,
+    1 => U1,
+    2 => U2,
+    3 => U3,
+    4 => U4,
+    5 => U5,
+    6 => U6,
+    7 => U7,
+    8 => U8,
+    9 => U9,
+    10 => U10,
+    11 => U11,
+    12 => U12,
+    13 => U13,
+    14 => U14,
+    15 => U15,
+    16 => U16,
+    17 => U17,
+    18 => U18,
+    19 => U19,
+    20 => U20,
+    21 => U21,
+    22 => U22,
+    23 => U23,
+    24 => U24,
+    25 => U25,
+    26 => U26,
+    27 => U27,
+    28 => U28,
+    29 => U29,
+    30 => U30,
+    31 => U31,
+    32 => U32,
+    33 => U33,
+    34 => U34,
+    35 => U35,
+    36 => U36,
+    37 => U37,
+    38 => U38,
+    39 => U39,
+    40 => U40,
+    41 => U41,
+    42 => U42,
+    43 => U43,
+    44 => U44,
+    45 => U45,
+    46 => U46,
+    47 => U47,
+    48 => U48,
+    49 => U49,
+    50 => U50,
+    51 => U51,
+    52 => U52,
+    53 => U53,
+    54 => U54,
+    55 => U55,
+    56 => U56,
+    57 => U57,
+    58 => U58,
+    59 => U59,
+    60 => U60,
+    61 => U61,
+    62 => U62,
+    63 => U63,
+    64 => U64,
+    96 => U96,
+    128 => U128,
+    192 => U192,
+    256 => U256,
+    384 => U384,
+    448 => U448,
+    512 => U512,
+    768 => U768,
+    896 => U896,
+    1024 => U1024,
+    2048 => U2048,
+    4096 => U4096,
+    8192 => U8192
+}
+
+/// Flexible generic array type.
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub struct Array<T, U: ArrayLength<T>>(pub U::ArrayType);
```

Comment:
`ArrayOps` is not implemented for all `Array`s, only for specific `<T, length>`s because of the macro; is it possible to have `as_slice` or similar methods on the struct, not just the `ArrayOps` trait? This is like what generic-array does.

Since the compiler doesn't know what type has `ArrayOps` implemented for it yet, it says it's not found. And all the other traits impl'd on `Array` like `AsRef` place a bound that requires `ArrayOps` to be impl'd for `Array`.

To get around this I impl'd an `ArrayExt`, but maybe we can add it to the `Array` struct?

```suggestion
pub struct Array<T, U: ArrayLength<T>>(pub U::ArrayType);

impl<T, U> Array<T, U>
where
    U: ArrayLength<T>,
{
    pub fn as_slice(&self) -> &[T] {
        self.0.as_ref()
    }
}
```